src/common/components/HatsProtocol/ConnectFarcasterAccountViaHatsProtocol.tsx

```diff
@@ -0,0 +1,631 @@
+    if (!address) return;
+
+    let hexStringPublicKey, hexStringPrivateKey;
+    if (
+      !hatsProtocolPendingAccounts ||
+      hatsProtocolPendingAccounts.length === 0
+    ) {
+      const { publicKey, privateKey } = await generateKeyPair();
+      hexStringPublicKey = bytesToHexString(publicKey)._unsafeUnwrap();
+      hexStringPrivateKey = bytesToHexString(privateKey)._unsafeUnwrap();
+
+      try {
+        addAccount({
+          id: null,
+          platformAccountId: fid.toString(),
+          status: AccountStatusType.pending,
+          platform: AccountPlatformType.farcaster_hats_protocol,
+          publicKey: hexStringPublicKey,
+          privateKey: hexStringPrivateKey,
+          data: {},
+        });
+      } catch (e) {
+        console.log("error when trying to add account", e);
+        setState(HatsProtocolSignupSteps[5]);
+        return;
+      }
+    } else {
+      hexStringPublicKey = hatsProtocolPendingAccounts[0].publicKey;
+      hexStringPrivateKey = hatsProtocolPendingAccounts[0].privateKey;
+    }
+
+    // const addressNounce = await readNoncesFromKeyGateway(address)
+    // console.log('KeyGateway: wallet nonce', addressNounce, 'delegatorContract nonce', nonce)
+    // const nonce = await readNoncesFromKeyGateway(address);
+    const nonce = await readNoncesFromKeyGateway(delegatorContractAddress);
+    const typedMetadataData = {
+      domain: SIGNED_KEY_REQUEST_VALIDATOR_EIP_712_DOMAIN,
+      types: {
+        SignedKeyRequest: SIGNED_KEY_REQUEST_TYPE,
+      },
+      primaryType: "SignedKeyRequest" as const,
+      message: {
+        requestFid: BigInt(fid),
+        key: hexStringPublicKey,
+        deadline: BigInt(deadline),
+      },
+    };
+    const metadataHash = hashTypedData(typedMetadataData);
+    const metadataSignature = await signTypedDataAsync(typedMetadataData);
+    console.log('metadataSignature', metadataSignature)
+
+    const METADATA_TYPEHASH = '0x16be47f1f1f50a66a48db64eba3fd35c21439c23622e513aab5b902018aec438';
+    const metadata = encodePacked(
+      ["bytes", "bytes32", "uint256", "bytes", "uint256"],
+      [metadataSignature, METADATA_TYPEHASH, BigInt(fid), keccak256(hexStringPublicKey), deadline]
+    );
+
+    console.log('isMetadataSignatureValid', await isValidSignature(metadataHash, metadata));
+
+    const typedData = {
+      domain: KEY_GATEWAY_EIP_712_DOMAIN,
+      types: {
+        Add: KEY_GATEWAY_ADD_TYPE,
+      },
+      primaryType: "Add" as const,
+      message: {
+        owner: delegatorContractAddress,
+        keyType: 1, // only supports 1 as key type
+        key: hexStringPublicKey,
+        metadataType: 1, // only supports 1 as metadata type
+        metadata,
+        deadline,
+        nonce,
+      },
+    };
+
+    const addHash = hashTypedData(typedData);
+    const addSignature = await signTypedDataAsync(typedData);
+
+    // signature, typehash, and then all the args
+    const sig = encodePacked(
```

Comment:
same here as with the `metadata` encoding: the last 8 values should be `encode()`-concatenated together, and the result should be `encodePacked()` with the `addSignature`